packages/headless/src/api/generated-answer/generated-answer-client.ts

```diff
@@ -0,0 +1,152 @@
+import {
+  GeneratedAnswerCitationsPayload,
+  GeneratedAnswerMessagePayload,
+  GeneratedAnswerPayloadType,
+  GeneratedAnswerStreamEventData,
+  GeneratedAnswerStreamFinishReason,
+} from './generated-answer-event-payload';
+import {GeneratedAnswerStreamRequest} from './generated-answer-request';
+
+export interface GeneratedAnswerAPIClientOptions {
+  logger: Logger;
+}
+
+export interface AsyncThunkGeneratedAnswerOptions<
+  T extends Partial<SearchAppState>
+> extends AsyncThunkOptions<
+    T,
+    ClientThunkExtraArguments<SearchAPIClient, GeneratedAnswerAPIClient>
+  > {}
+
+const buildStreamingUrl = (url: string, orgId: string, streamId: string) =>
+  new URLPath(
+    `${url}/rest/organizations/${orgId}/machinelearning/streaming/${streamId}`
+  );
+
+const MAX_RETRIES = 3;
+const MAX_TIMEOUT = 5000;
+export const RETRYABLE_STREAM_ERROR_CODE = 1;
+
+export class GeneratedAnswerAPIClient {
+  private logger: Logger;
+
+  constructor(private options: GeneratedAnswerAPIClientOptions) {
+    this.logger = options.logger;
+  }
+
+  streamGeneratedAnswer(
+    params: GeneratedAnswerStreamRequest,
+    onMessage: (payload: GeneratedAnswerMessagePayload) => void,
+    onCitations: (payload: GeneratedAnswerCitationsPayload) => void,
+    onError: (payload: SSEErrorPayload) => void,
+    onCompleted: () => void
+  ) {
+    const {url, organizationId, streamId, accessToken} = params;
+
+    if (!streamId) {
+      this.logger.error('No stream ID found');
+      return;
+    }
+
+    let retryCount = 0;
+    let timeout: ReturnType<typeof setTimeout>;
+    let source: EventSourcePolyfill;
+
+    const checkAndRetry = (e: unknown): EventSourcePolyfill | void => {
+      const errorMessage = 'Failed to connect to stream.';
+      this.logger.error(errorMessage, e);
+
+      if (++retryCount <= MAX_RETRIES) {
+        this.logger.info(`Retrying...(${retryCount}/${MAX_RETRIES})`);
+        return stream();
+      } else {
+        this.logger.info('Maximum retry exceeded.');
+        onError({
+          message: errorMessage,
+          code: RETRYABLE_STREAM_ERROR_CODE,
+        });
+      }
+    };
+
+    const refreshTimeout = () => {
+      timeout = resetTimeout(timeout, checkAndRetry, MAX_TIMEOUT);
+    };
+
+    const handleStreamPayload = (
+      payloadType: GeneratedAnswerPayloadType,
+      payload: string
+    ) => {
+      if (payloadType === GeneratedAnswerPayloadType.Message) {
+        onMessage(JSON.parse(payload) as GeneratedAnswerMessagePayload);
+      } else if (payloadType === GeneratedAnswerPayloadType.Citations) {
+        onCitations(JSON.parse(payload) as GeneratedAnswerCitationsPayload);
+      }
+    };
+
+    const stream = (): EventSourcePolyfill | void => {
+      try {
+        source = new EventSourcePolyfill(
+          buildStreamingUrl(url, organizationId, streamId).href,
+          {
+            headers: {
+              Authorization: `Bearer ${accessToken}`,
```

Comment:
The stream API has a way to know that a given access token has access to retrieve a particular stream ID ?

Do you know how that works ?

Reply:
The stream API does indeed check access. For now it only checks that the user has query_execute rights but in a few weeks it will also make sure that the user reading the stream is the same one that generated it.